libs/checkpoint/src/store/postgres.ts

```diff
@@ -0,0 +1,714 @@
+              params.push(key, JSON.stringify(value));
+              paramIndex += 2;
+            }
+          }
+          if (filterConditions.length > 0) {
+            query += ` AND ${filterConditions.join(" AND ")}`;
+          }
+        }
+
+        // Vector search
+        if (op.query && this.indexConfig?.embeddings) {
+          const embedding = await this.indexConfig.embeddings.embedQuery(op.query);
+
+          const vectorParamIndex = paramIndex;  // Store the index for the vector parameter
+          query = `
+            WITH scored AS (
+              SELECT s.prefix, s.key, s.value, s.created_at, s.updated_at,
+                      1 - (sv.embedding <=> $${vectorParamIndex}::vector) as score
+              FROM store s
+              JOIN store_vectors sv ON s.prefix = sv.prefix AND s.key = sv.key
+              WHERE s.prefix LIKE $1
+              ${filterConditions.length > 0 ? `AND ${filterConditions.join(" AND ")}` : ""}
+              ORDER BY score DESC
+              LIMIT $${vectorParamIndex + 1}
+              OFFSET $${vectorParamIndex + 2}
+            )
+            SELECT DISTINCT ON (prefix, key)
+              prefix, key, value, created_at, updated_at, score
+            FROM scored
+            ORDER BY prefix, key, score DESC
+          `;
+          
+          params.push(JSON.stringify(embedding));      // $vectorParamIndex
+          params.push(op.limit ?? 10); // $vectorParamIndex + 1
+          params.push(op.offset ?? 0); // $vectorParamIndex + 2
+        } else {
+          // Regular search
+          query += ` ORDER BY updated_at DESC LIMIT $${paramIndex} OFFSET $${
+            paramIndex + 1
+          }`;
+          params.push(op.limit ?? 10);
+          params.push(op.offset ?? 0);
+        }
+
+        const { rows } = await client.query<Row>(query, params);
+        this.updateResults(
+          results, 
+          idx, 
+          rows.map(row => this.createSearchResult(row))
+        );
+      }
+    } finally {
+      client.release();
+    }
+  }
+
+  private getFilterCondition(
+    key: string,
+    op: string,
+    value: unknown,
+    startIndex: number
+  ): [string, unknown[]] {
+    switch (op) {
+      case "$eq":
+        return [`value->$${startIndex} = $${startIndex + 1}::jsonb`, [key, JSON.stringify(value)]];
+      case "$gt":
+        return [`value->>$${startIndex} > $${startIndex + 1}`, [key, String(value)]];
+      case "$gte":
+        return [`value->>$${startIndex} >= $${startIndex + 1}`, [key, String(value)]];
+      case "$lt":
+        return [`value->>$${startIndex} < $${startIndex + 1}`, [key, String(value)]];
+      case "$lte":
+        return [`value->>$${startIndex} <= $${startIndex + 1}`, [key, String(value)]];
+      case "$ne":
+        return [`value->$${startIndex} != $${startIndex + 1}::jsonb`, [key, JSON.stringify(value)]];
+      default:
+        throw new Error(`Unsupported operator: ${op}`);
```

Comment:
Am I correct that the arguments to this function may have been user input? If that's the case, you'll need to put `startIndex` through a `parseInt` to validate it, otherwise it's a potential SQL injection vulnerability.